### PR TITLE
Implementation multiplayer readiness.

### PR DESCRIPTION
**Pull request by o01eg**
- Add the field to the game protocol and lobby UI.
- Drop it if the important game's or players' changes were happened.
- When all human players are ready the game starts without requiring host player to press "start" button.

### CODE REVIEW

File: UI/MultiplayerLobbyWnd.cpp [hunk 1]
```diff
@@ -229,6 +229,8 @@ namespace {
                 push_back(new CUILabel(""));
                 push_back(new CUILabel(""));
                 push_back(new CUILabel(""));
+                push_back(new CUILabel(UserString(player_data.m_player_ready ? "YES" : "NO")));
```
Comment (adrianbroher) [inline UI/MultiplayerLobbyWnd.cpp]: Only use string literals or variables as `UserString` parameter.

File: UI/MultiplayerLobbyWnd.cpp [hunk 2]
```diff
@@ -260,6 +262,12 @@ namespace {
                 species_selector->Disable();
             else
                 GG::Connect(species_selector->SpeciesChangedSignal, &NewGamePlayerRow::SpeciesChanged,      this);
+
+            // ready check box or empty label for AI
+            if (player_data.m_client_type == Networking::CLIENT_TYPE_AI_PLAYER)
+                push_back(new CUILabel(""));
+            else
+                push_back(new CUILabel(UserString(player_data.m_player_ready ? "YES" : "NO")));
```
Comment (adrianbroher) [inline UI/MultiplayerLobbyWnd.cpp]: Only use string literals or variables as `UserString` parameter.

File: UI/MultiplayerLobbyWnd.cpp [hunk 3]
```diff
@@ -342,6 +350,12 @@ namespace {
                 m_empire_list->Disable();
             else
                 Connect(m_empire_list->SelChangedSignal, &LoadGamePlayerRow::EmpireChanged, this);
+
+            // ready check box or empty label for AI
+            if (player_data.m_client_type == Networking::CLIENT_TYPE_AI_PLAYER)
+                push_back(new CUILabel(""));
+            else
+                push_back(new CUILabel(UserString(player_data.m_player_ready ? "YES" : "NO")));
```
Comment (adrianbroher) [inline UI/MultiplayerLobbyWnd.cpp]: Only use string literals or variables as `UserString` parameter.

**Review by adrianbroher — CHANGES_REQUESTED** [review]
When using `UserString` or relatives you either wrap a string literal or a local variable.  If you can't use a string literal at the place where the translation occurs use `UserStringNop` to wrap the string literal to be translated.

**Comment by o01eg** [comment]
Fixed.

**Review by adrianbroher — APPROVED** [review]

**Comment by adrianbroher** [comment]
The review only covered the incorrect usage of `UserString`.  No further inspection was done.

File: UI/MultiplayerLobbyWnd.cpp [hunk 4]
```diff
@@ -543,10 +561,9 @@ void MultiPlayerLobbyWnd::KeyPress(GG::Key key, boost::uint32_t key_code_point,
         // put message just sent in chat box (local echo)
         *m_chat_box += player_name + ": " + text + "\n";
 
-    } else if (m_start_game_bn && !m_start_game_bn->Disabled() &&
-               (key == GG::GGK_RETURN || key == GG::GGK_KP_ENTER))
+    } else if (m_ready_bn && (key == GG::GGK_RETURN || key == GG::GGK_KP_ENTER))
```
Comment (dbenage-cx) [inline UI/MultiplayerLobbyWnd.cpp]: Bring opening brace up to this line
`} else if () {`

File: UI/MultiplayerLobbyWnd.cpp [hunk 5]
```diff
@@ -676,10 +695,10 @@ void MultiPlayerLobbyWnd::DoLayout() {
...
     m_players_lb->SizeMove(players_lb_ul, players_lb_lr);
 
-    m_start_game_bn->SizeMove(GG::Pt(GG::X0, GG::Y0), GG::Pt(GG::X(125), m_start_game_bn->MinUsableSize().y));
-    m_cancel_bn->SizeMove(GG::Pt(GG::X0, GG::Y0), GG::Pt(GG::X(125), m_start_game_bn->MinUsableSize().y));
+    m_ready_bn->SizeMove(GG::Pt(GG::X0, GG::Y0), GG::Pt(GG::X(125), m_ready_bn->MinUsableSize().y));
+    m_cancel_bn->SizeMove(GG::Pt(GG::X0, GG::Y0), GG::Pt(GG::X(125), m_ready_bn->MinUsableSize().y));
     m_cancel_bn->MoveTo(GG::Pt(ClientWidth() - m_cancel_bn->Width() - CONTROL_MARGIN, ClientHeight() - m_cancel_bn->Height() - CONTROL_MARGIN));
-    m_start_game_bn->MoveTo(GG::Pt(m_cancel_bn->RelativeUpperLeft().x - CONTROL_MARGIN - m_start_game_bn->Width(),
+    m_ready_bn->MoveTo(GG::Pt(m_cancel_bn->RelativeUpperLeft().x - CONTROL_MARGIN - m_ready_bn->Width(),
                                    ClientHeight() - m_cancel_bn->Height() - CONTROL_MARGIN));
```
Comment (dbenage-cx) [inline UI/MultiplayerLobbyWnd.cpp]: Align with start of parameter list (or indented one level for long lines)
`ClientHeight` should align with `m_cancel_bn` in this case

File: UI/MultiplayerLobbyWnd.cpp [hunk 6]
```diff
@@ -787,6 +808,15 @@ bool MultiPlayerLobbyWnd::PopulatePlayerList() {
...
         }
+
+        // checks for ready button
+        if (data_player_id == HumanClientApp::GetApp()->PlayerID())
+            is_client_ready = psd.m_player_ready;
+        else if (psd.m_client_type == Networking::CLIENT_TYPE_HUMAN_PLAYER ||
+            psd.m_client_type == Networking::CLIENT_TYPE_HUMAN_OBSERVER ||
+            psd.m_client_type == Networking::CLIENT_TYPE_HUMAN_MODERATOR) {
+            is_other_ready = is_other_ready && psd.m_player_ready;
+        }
```
Comment (dbenage-cx) [inline UI/MultiplayerLobbyWnd.cpp]: Multi-line should start brace non-indented on next line.
```
else if (b ||
    (c && d))
{ x; }

File: server/ServerFSM.cpp [hunk 7]
```diff
@@ -113,6 +113,7 @@ namespace {
             default:                                        ss << "<invalid client type>, ";
             }
             ss << it->second.m_starting_species_name;
+            if (it->second.m_player_ready) { ss << ", Ready"; }
```
Comment (dbenage-cx) [inline server/ServerFSM.cpp]: body to next line (can drop braces)
```
if (a)
    x;

File: server/ServerFSM.cpp [hunk 8]
```diff
@@ -506,6 +523,7 @@ sc::result MPLobby::react(const LobbyUpdate& msg) {
     // check if new lobby data changed player setup data.  if so, need to echo
     // back to sender with updated lobby details.
     bool player_setup_data_changed = (incoming_lobby_data.m_players != m_lobby_data->m_players);
+    bool is_drop_ready = false;
```
Comment (dbenage-cx) [inline server/ServerFSM.cpp]: `is_drop_ready` is unclear in this context, prefer (e.g.) `galaxy_setup_data_changed` or `lobby_data_changed`

File: server/ServerFSM.cpp [hunk 9]
```diff
@@ -685,6 +737,34 @@ sc::result MPLobby::react(const LobbyUpdate& msg) {
...
+    if (is_drop_ready) {
+        for (std::pair<int, PlayerSetupData>& player : m_lobby_data->m_players)
+            player.second.m_player_ready = false;
+    } else {
+        // check if all established human players ready to play
+        bool is_all_ready = true;
+        for (std::pair<int, PlayerSetupData>& player : m_lobby_data->m_players) {
+            if ((player.first >= 0) && (player.second.m_client_type == Networking::CLIENT_TYPE_HUMAN_OBSERVER ||
+                player.second.m_client_type == Networking::CLIENT_TYPE_HUMAN_MODERATOR ||
+                player.second.m_client_type == Networking::CLIENT_TYPE_HUMAN_PLAYER)) {
```
Comment (dbenage-cx) [inline server/ServerFSM.cpp]: multi-line brace to next line

**Review by dbenage-cx — CHANGES_REQUESTED** [review]

**Comment by dbenage-cx** [comment]
Still needs adjustment for (at least) label layout.
Something is off as every time I nudge the position a little, the labels end up far offset.  Will try to spend some time looking into it this week if noone jumps in before.

**Comment by o01eg** [comment]
I've fixed code style and some bugs.

**Review by dbenage-cx — APPROVED** [review]

**Comment by Vezzra** [comment]
Has anyone (besides @o01eg) actually ran some tests if the PR works as intended?

If yes, and given that we have approved code reviews, LGTM I guess...?

**Comment by dbenage-cx** [comment]
@Vezzra still needs UI work and then thorough testing

**Comment by Vezzra** [comment]
Ok.

**Comment by o01eg** [comment]
I've add commits from dbenage-cx with reworked UI that looks more better.

File: UI/MultiplayerLobbyWnd.cpp [hunk 10]
```diff
@@ -37,6 +37,14 @@ namespace {
     { return ClientUI::GetFont(ClientUI::Pts())->Height(); }
     GG::Y PlayerRowHeight()
     { return PlayerFontHeight() + 2 * PlayerRowMargin(); }
+    GG::X PlayerReadyBrowseWidth()
+    { return GG::X(ClientUI::Pts() * 11); }
+
+    const boost::shared_ptr<GG::Texture> GetReadyTexture(bool ready) {
+        if (ready)
+            return GG::GetTextureManager().GetTexture(ClientUI::ArtDir() / "icons/ready.png");
```
Comment (geoffthemedio) [inline UI/MultiplayerLobbyWnd.cpp]: GG::GetTextureManager seems to not be publically accessible to my linker. Replace this with standard in FreeOrion UI code:
````
ClientUI::GetTexture(...)
````

**Review by geoffthemedio — CHANGES_REQUESTED** [review]

**Comment by geoffthemedio** [comment]
Column headers look nice, though I'd rather other similar lists looked the same elsehwere in the UI.

New icon based indicators are good; was going to request something similar.

It looks like I should be getting tooltips / browse windows for something, thought I don't see any in-game...

File: UI/MultiplayerLobbyWnd.cpp (hunk 10 above)
Comment (dbenage-cx) [reply UI/MultiplayerLobbyWnd.cpp]: Fixed (via web interface)

**Comment by dbenage-cx** [comment]
I may have missed something with tooltips, but I suspect `Row` is simply not passing the pointer through.
Will continue to investigate after other work.

**Comment by dbenage-cx** [comment]
4cf9661 Fixes tooltip display
2b0a27f Shifts the status to first column (if desired, no big issue with using the last column)

**Comment by geoffthemedio** [comment]
Why leftmost and not rightmost column?

**Comment by o01eg** [comment]
I can remove last commit if it not neeeded.

**Comment by geoffthemedio** [comment]
last column seemed fine to me.

**Comment by o01eg** [comment]
Ok.

**Comment by geoffthemedio** [comment]
I find the current button labels or design unclear. It's not obvious whether clicking a button that says "Ready" would indicate that you are ready, or that it would toggle back to "Not Ready". The "Start Game" button is better, as it's clearly a verb, so perhaps "Become Ready" or "Declare Ready" or "Accept"... and something similar for the undo buttons?

Also, some indication in the list which player is the host, and can thus edit other players' lobby selections, would be good, as it's no so obvious now due to anyone being able to click ready/not ready.

**Comment by o01eg** [comment]
I've rebased code and added new column for host but cann't make it appears without scrolling.

**Review by geoffthemedio — APPROVED** [review]

**Comment by o01eg** [comment]
It's not yet ready due UI issues.

**Comment by dbenage-cx** [comment]
If a new column is to be added, we need to increase the default width as the controls are too small already.  (Not certain there is much loss wrt who the host is, except in the case of moderators.)

Other options:
*  Add the host icon into the ready state column (requires a change to `MultiTextureStaticGraphic`).
*  Add such an icon behind the player name.
*  Have some indicator (external to this list) which denotes to host, possibly with other info.

File: UI/MultiplayerLobbyWnd.cpp [hunk 11]
```diff
@@ -462,6 +493,7 @@ namespace {
         retval.push_back(color_width); // color
         retval.push_back(prop_width); // species/original player
         retval.push_back(ready_width); // player ready
+        retval.push_back(ready_width); // host
```
Comment (dbenage-cx) [inline UI/MultiplayerLobbyWnd.cpp:496]: Adjust `prop_width` to account for new fixed width column.

Comment (o01eg) [reply UI/MultiplayerLobbyWnd.cpp:496]: Thanks, now it looks correctly.

**Comment by o01eg** [comment]
I've fixed UI.

**Comment by Vezzra** [comment]
Can the "work in progress" label be removed, that is, is the PR ready for final review and merge? Or is there still some work/adjustments that need to be done?

**Comment by o01eg** [comment]
No, I think it almost done.

**Comment by geoffthemedio** [comment]
I'll merge now, though I'd like a tooltip for the host indicator column, and rephrased tooltips for "decline" and "accept", with something like "setup accepted" and "setup not yet accepted"

**Pull request merged**